### PR TITLE
Jsonnet type system

### PR DESCRIPTION
**Pull request by CertainLach**

### CODE REVIEW

File: crates/jrsonnet-types/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		];
+		ComplexValType::UnionRef(CONTENTS)
+	}};
+	(($($a:tt) &+)) => {{
+		static CONTENTS: &'static [ComplexValType] = &[
+			$(ty!($a)),+
+		];
+		ComplexValType::SumRef(CONTENTS)
+	}};
+}
```
Comment (JarvisCraft) [inline crates/jrsonnet-types/src/lib.rs:51]: Sholdn't absolute `$crate`-based paths be used for an exported macro?

File: bindings/jsonnet/src/val_make.rs [hunk 2]
```diff
@@ -38,7 +37,7 @@ pub extern "C" fn jsonnet_json_make_null(_vm: &EvaluationState) -> *mut Val {
 
 #[no_mangle]
 pub extern "C" fn jsonnet_json_make_array(_vm: &EvaluationState) -> *mut Val {
-	Box::into_raw(Box::new(Val::Arr(Rc::new(Vec::new()))))
+	todo!()
```
Comment (JarvisCraft) [inline bindings/jsonnet/src/val_make.rs:40]: PR should be converted to draft

File: bindings/jsonnet/src/val_modify.rs [hunk 3]
```diff
@@ -12,17 +12,10 @@ use std::{collections::HashMap, ffi::CStr, os::raw::c_char, rc::Rc};
...
 ) {
-	match *Box::from_raw(arr) {
-		Val::Arr(old) => {
-			let mut new = Rc::try_unwrap(old).expect("arr with no refs");
-			new.push(val.clone());
-			*arr = Val::Arr(Rc::new(new));
-		}
-		_ => panic!("should receive array"),
-	}
+	todo!()
```
Comment (JarvisCraft) [inline bindings/jsonnet/src/val_modify.rs:18]: PR should be converted to draft

File: crates/jrsonnet-types/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+[package]
+name = "jrsonnet-types"
+version = "0.3.2"
+authors = ["Yaroslav Bolyukin <iam@lach.pw>"]
```
Comment (JarvisCraft) [inline crates/jrsonnet-types/Cargo.toml:4]: Consider using the same style for all `Cargo.toml`s (see #28)

**Review by JarvisCraft — CHANGES_REQUESTED** [review]

File: crates/jrsonnet-types/src/lib.rs (hunk 1 above)
Comment (CertainLach) [reply crates/jrsonnet-types/src/lib.rs:51]: Yep, it should, but tgis is temporary syntax
I want to write proc macro instead for parsing typescript-like type annotations

Comment (CertainLach) [reply crates/jrsonnet-types/src/lib.rs:51]: Proc macro would be too much, leaving it as is (With paths moved to $crate)

**Comment by CertainLach** [comment]
Because of array access optimizations, this branch is even faster than master (0.3.3):
```
Benchmark #1: ./master -J vendor -m manifests example.jsonnet
  Time (mean ± σ):     344.7 ms ±   4.5 ms    [User: 292.8 ms, System: 51.9 ms]
  Range (min … max):   339.8 ms … 355.0 ms    10 runs
 
Benchmark #2: ./type-safety -J vendor -m manifests example.jsonnet
  Time (mean ± σ):     298.9 ms ±   4.5 ms    [User: 258.8 ms, System: 40.0 ms]
  Range (min … max):   293.4 ms … 308.1 ms    10 runs
 
Summary
  './type-safety -J vendor -m manifests example.jsonnet' ran
    1.15 ± 0.02 times faster than './master -J vendor -m manifests example.jsonnet'
```

(Tested on https://github.com/prometheus-operator/kube-prometheus)

**Comment by CertainLach** [comment]
Well, i think type descriptions on syntax level should be out of scope of this pullrequest

**Pull request merged**